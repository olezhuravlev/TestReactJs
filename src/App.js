import './App.css';
import React, {useEffect, useState} from "react";
import {BrowserRouter, Link, Route, Routes} from "react-router-dom";
import MainPage from "./pages/MainPage";
import AboutPage from "./pages/AboutPage";

function App() {

    const [data, setData] = useState(null);
    const [toggle, setToggle] = useState(false);
    const [value, setValue] = useState('');

    const onClick = () => {
        setToggle(prevState => !prevState);
    }

    useEffect(() => {
        setTimeout(() => {
            console.log('Delayed action!')
            setData({});
        }, 1000);
    }, []);

    return (
        <BrowserRouter>
            <Link to='' data-testid='main-link-testid'>Main</Link>
            <Link to='/about' data-testid='about-link-testid'>About</Link>
            <Routes>
                <Route path={'/'} element={<MainPage/>}/>
                <Route path={'/about'} element={<AboutPage/>}/>
            </Routes>
            <div>
                <div>
                    <h1>Hello world!</h1>
                    <button style={{color: 'red'}} className={"buttonClass"} onClick={onClick} data-testid={"button-test-id"}>Click!
                    </button>
                    <input type={"text"} placeholder={"input value..."} data-testid={"input-test-id"}
                           onChange={(e) => setValue(e.target.value)}/>
                    <h2 data-testid={"text-test-id"}>{value}</h2>
                    {data && <div>Data ready!</div>}
                    {toggle && <div data-testid={"toggle-div-test-id"}>Button toggled!</div>}
                </div>
            </div>
        </BrowserRouter>
    );
}

export default App;
