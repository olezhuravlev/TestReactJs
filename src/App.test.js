import {fireEvent, render, screen} from '@testing-library/react';
import App from './App';
import userEvent from "@testing-library/user-event";

describe("Test my-app", () => {

    test('Widgets presence', () => {

        render(<App/>);

        const hello = screen.getByText(/hello world/i);
        expect(hello).toBeInTheDocument();
        expect(hello).toMatchSnapshot();

        const button = screen.getByRole("button");
        expect(button).toBeInTheDocument();
        expect(button).toMatchSnapshot();

        const input = screen.getByPlaceholderText(/input value/i);
        expect(input).toBeInTheDocument();
        expect(input).toMatchSnapshot();

        screen.debug();
    });

    test('Delayed widgets presence', async () => {

        render(<App/>);

        screen.debug();
        const dataElement = await screen.findByText(/data ready/i);
        expect(dataElement).toBeInTheDocument();
        screen.debug();

        const button = screen.getByRole("button");
        expect(button).toHaveStyle({color: 'red'});
    });

    it('CLICK EVENT', () => {

        render(<App/>);

        const button = screen.getByTestId("button-test-id");
        expect(button).toBeInTheDocument();

        expect(screen.queryByTestId("toggle-div-test-id")).toBeNull();
        fireEvent.click(button);
        expect(screen.queryByTestId("toggle-div-test-id")).toBeInTheDocument();
    })

    it('INPUT EVENT with fireEvent', () => {

        render(<App/>);

        const input = screen.getByTestId('input-test-id');
        expect(input).toBeInTheDocument();

        const h2 = screen.getByTestId('text-test-id');
        expect(h2).toContainHTML('');
        expect(h2).toHaveTextContent('');

        const testText = 'test input value';
        fireEvent.input(input, {
            target: {value: testText}
        });
        expect(h2).toContainHTML(testText);
        expect(h2).toHaveTextContent(testText);

        const text = screen.getByTestId("text-test-id");
        expect(text).toBeInTheDocument();
    })

    it('INPUT EVENT with userEvent', () => {

        render(<App/>);

        const input = screen.getByTestId('input-test-id');
        expect(input).toBeInTheDocument();

        const h2 = screen.getByTestId('text-test-id');
        expect(h2).toContainHTML('');
        expect(h2).toHaveTextContent('');

        const testText = 'test input value';
        userEvent.type(input, testText)
        expect(h2).toContainHTML(testText);
        expect(h2).toHaveTextContent(testText);

        const text = screen.getByTestId("text-test-id");
        expect(text).toBeInTheDocument();
    })
})
