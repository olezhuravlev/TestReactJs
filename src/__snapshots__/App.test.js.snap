// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test my-app Widgets presence 1`] = `
<h1>
  Hello world!
</h1>
`;

exports[`Test my-app Widgets presence 2`] = `
<button
  class="buttonClass"
  data-testid="button-test-id"
  style="color: red;"
>
  Click!
</button>
`;

exports[`Test my-app Widgets presence 3`] = `
<input
  data-testid="input-test-id"
  placeholder="input value..."
  type="text"
/>
`;
